image_proc/src/nodelets/resize.cpp

```diff
@@ -1,171 +1,183 @@
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+#include <mutex>
 #include <cv_bridge/cv_bridge.h>
 #include <dynamic_reconfigure/server.h>
 #include <nodelet/nodelet.h>
-#include <nodelet_topic_tools/nodelet_lazy.h>
 #include <ros/ros.h>
 #include <sensor_msgs/CameraInfo.h>
 #include <sensor_msgs/Image.h>
+#include <image_transport/image_transport.h>
 
 #include "image_proc/ResizeConfig.h"
 
 namespace image_proc
 {
-
-class ResizeNodelet : public nodelet_topic_tools::NodeletLazy
+class ResizeNodelet : public nodelet::Nodelet
 {
 protected:
   // ROS communication
-  ros::Publisher pub_image_;
-  ros::Publisher pub_info_;
-  ros::Subscriber sub_info_;
-  ros::Subscriber sub_image_;
+  image_transport::CameraPublisher pub_image_;
+  image_transport::CameraSubscriber sub_image_;
+  int queue_size_;
+
+  std::shared_ptr<image_transport::ImageTransport> it_, private_it_;
+  std::mutex connect_mutex_;
 
   // Dynamic reconfigure
-  boost::recursive_mutex config_mutex_;
+  std::mutex config_mutex_;
   typedef image_proc::ResizeConfig Config;
   typedef dynamic_reconfigure::Server<Config> ReconfigureServer;
-  boost::shared_ptr<ReconfigureServer> reconfigure_server_;
+  std::shared_ptr<ReconfigureServer> reconfigure_server_;
   Config config_;
 
   virtual void onInit();
-  virtual void subscribe();
-  virtual void unsubscribe();
 
-  void imageCb(const sensor_msgs::ImageConstPtr& image_msg);
+  void connectCb();
+
+  void imageCb(const sensor_msgs::ImageConstPtr& image_msg, const sensor_msgs::CameraInfoConstPtr& info_msg);
   void infoCb(const sensor_msgs::CameraInfoConstPtr& info_msg);
 
-  void configCb(Config &config, uint32_t level);
+  void configCb(Config& config, uint32_t level);
 };
 
 void ResizeNodelet::onInit()
 {
-  nodelet_topic_tools::NodeletLazy::onInit();
+  ros::NodeHandle& nh = getNodeHandle();
+  ros::NodeHandle& private_nh = getPrivateNodeHandle();
+  it_.reset(new image_transport::ImageTransport(nh));
+  private_it_.reset(new image_transport::ImageTransport(private_nh));
```

Comment:
I'm not super familiar with the implementation differences between a regular subscriber and an ImageTransport subscriber. Is there a reason that this one uses the private node handle?

Reply:
When I use the regular node handle I get errors like this:
```bash
[ERROR] [1556808041.833527432]: Tried to advertise a service that is already advertised in this node [/camera/depth/image_rect_raw/compressedDepth/set_parameters]
[ERROR] [1556808041.837569377]: Tried to advertise a service that is already advertised in this node [/camera/depth/image_rect_raw/compressed/set_parameters]
[ERROR] [1556808041.841201998]: Tried to advertise a service that is already advertised in this node [/camera/depth/image_rect_raw/theora/set_parameters]
```
I think the source of this Error is located in the in image_transport_plugins library. The CompressedPublisher and the CompressedSubscriber are initialized with the same base topic so their dynamic_reconfigure services collide. By using the private namespace for the CameraPublisher this issue is avoided.

Reply:
Check out https://github.com/Tuebel/image_pipeline/pull/1

Reply:
The image transport framework should use private nodehandles for the reasons above. It exposes the available image_transport_plugins for compression under the main namespace to offer `*/compressed` `*/theora` etc so if you dont make them private, they'll collide, but for organizational purposes they won't belong to the sensor namespace they are representing